### ISSUE TITLE
Set  Time Out duration for requested association for C-Echo / C-Store

### ISSUE THREAD

**Issue by mohantyytasre** [fo-dicom/fo-dicom#1752]
### Expected behavior 
 To allow the client to set the association requested time duration if association fails and to abort the association and client.

I am using  C-Echo/C-store SCP program codes to test the Server with particular AET, IP address and Port #.
After providing wrong AET/IP address / Port # for server association request must get terminated after Association Request Timed Outset duration .

### Actual behavior
I am not able to set the association requested time duration.
Please help me with this for implementation as I am still learning c# and so not able to get the syntax

### Steps to reproduce the behavior
DicomCEchoRequest   CEchoRequest    = new DicomCEchoRequest();
            
            _echoServerHost = store_ServerHost;
            _echoServerPort = store_ServerPort;
            _echoServerAET = called_aet;
            _echoClientAET = calling_aet;

            var client = CreateDicomClient(store_ServerHost, store_ServerPort, calling_aet, called_aet);
            
            client.AssociationRequestTimedOut += (sender, e) =>
            {
                var t = e.TimeoutInMs;
               UpdateServerStatus("Association accepted by SCP");
            };

### fo-dicom version and OS/platform
fo-dicom : 5.1.1
OS : Windows

**Comment by ChrisNewell-Inari** [comment]
In the API Documentation, the [DicomClient](https://fo-dicom.github.io/stable/v5/api/FellowOakDicom.Network.Client.DicomClient.html#FellowOakDicom_Network_Client_DicomClient__ctor_System_String_System_Int32_ITlsInitiator_System_String_System_String_FellowOakDicom_Network_Client_DicomClientOptions_FellowOakDicom_Network_DicomServiceOptions_ILoggerFactory_IAdvancedDicomClientConnectionFactory_) constructonr lets you set the AssociationRequestTimeoutInMs via the [DicomClientOptions](https://fo-dicom.github.io/stable/v5/api/FellowOakDicom.Network.Client.DicomClientOptions.html) hope this helps!

**Comment by mohantyytasre** [comment]
Thanks for reply....
I did used it and found that default Association Request Time Out duration is 5000 ms. I have added the Eventhandler with client as   

            client.AssociationRequestTimedOut += (sender, e) =>
            {
                UpdateServerStatus($"Association requested for server time out: "+e.String());
            };

But even after no response from server after 5 sec. this event handler is not getting accessed by program. But also the system.exception error is getting logged as Exception thrown: 'System.Net.Sockets.SocketException' in System.dll.
So is it that where it is not allowing and exiting the current thread.
Also is there a way to cancel a request sent by client because in some discussions I have read that C-Cancel is not supported by fo-dicom.

**Comment by ChrisNewell-Inari** [comment]
Maybe A-Abort?

**Comment by gofal** [comment]
Take a look at this class: [Tests/FO-DICOM.Tests/Network/Client/DicomClientTimeoutTests.cs](https://github.com/fo-dicom/fo-dicom/blob/b45ffe1861d178b1f73aa048c8ae168fbd5f581b/Tests/FO-DICOM.Tests/Network/Client/DicomClientTimeoutTests.cs)
This are the unit tests about timeouts in clients. There you also find a method that tests the AssociationRequestTimeout: https://github.com/fo-dicom/fo-dicom/blob/b45ffe1861d178b1f73aa048c8ae168fbd5f581b/Tests/FO-DICOM.Tests/Network/Client/DicomClientTimeoutTests.cs#L620
Basically, you have to wrap your `await client.SendAsync()` in a try-catch-block and catch the DicomAssociationRequestTimeOutException instead of only adding a handler to the event. You should catch all exceptions, also the SocketException.

You also can cancel requests. You can pass a `CancellationToken` as parameter to `client.SendAsync(cancellationToken)`.

**Comment by jaime-olivares** [comment]
This one seems like an abandoned issue/question. No more feedback from the OP.

**Comment by gofal** [comment]
Closing this issue. If there is still a problem, then please reopen the issue and provide some actual informations.
